pipeline-runner/R/gem2s-2-load_user_files.R

```diff
@@ -263,3 +263,134 @@ format_annot <- function(annot_list) {
+    stop("Incompatible features detected.")
+  }
+
+  if (any(feature_types_list == IDS_SYM) &&
+      (any(feature_types_list == IDS_IDS) ||
+       any(feature_types_list == SYM_SYM))) {
+
+    annots_with_ids <-
+      unique(do.call("rbind", annot_list[feature_types_list == IDS_SYM]))
+
+    annots_with_ids <-
+      annots_with_ids[!duplicated(annots_with_ids$input),]
+
+    for (sample in samples) {
+      if (feature_types_list[[sample]] == SYM_SYM ||
+          feature_types_list[[sample]] == IDS_IDS) {
+        sample_annot <- annot_list[[sample]]
+
+        # Try to replace input column (currently symbols) in sample_annot with ids from annots_with_ids
+        if (feature_types_list[[sample]] == SYM_SYM) {
+          matched_symbols_index <-
+            match(sample_annot$input, annots_with_ids$name)
+          is_in_annot_list <-
+            which(sample_annot$input %in% annots_with_ids$name)
+
+          sample_annot$input[is_in_annot_list] <-
+            annots_with_ids$input[na.omit(matched_symbols_index)]
+
+          #This avoids duplicates after combining with the annotated df.
+          #Leads to a mismatch in genes between samples but it seems like the best solution
+          sample_annot$input <- make.unique(sample_annot$input)
+
+          #The counts have been loaded with gene symbols so we need to replace the rownames with the ids
+          rownames(counts_list[[sample]]) <- sample_annot$input
+        }
+
+        # Try to replace names column (currently ids) in sample_annot with symbols from annots_with_ids
+        if (feature_types_list[[sample]] == IDS_IDS) {
+          matched_ids_index <- match(sample_annot$input, annots_with_ids$input)
+          is_in_annot_list <-
+            which(sample_annot$input %in% annots_with_ids$input)
+
+          sample_annot$name[is_in_annot_list] <-
+            annots_with_ids$name[na.omit(matched_ids_index)]
+        }
+
+        annot_list[[sample]] <- sample_annot
+      }
+    }
+  }
+  return(list(counts_list=counts_list, annot_list=annot_list))
+}
+
+#' Determine the type of features in the annot data frame
+#'
+#' Classifies the features file columns into either ensemblIds or symbols
+#'
+#' @param annot data.frame read from features file
+#' @return character vector indicating feature types
+#'
+#' @export
+get_feature_types <- function(annot) {
+  is_ens_col1 <- startsWith(annot[[1]], "ENS")
+  pct_ens_col1 <- sum(is_ens_col1) / nrow(annot)
+
+  is_ens_col2 <- startsWith(annot[[2]], "ENS")
+  pct_ens_col2 <- sum(is_ens_col2) / nrow(annot)
+
+  is_ens <- c(pct_ens_col1 >= 0.5, pct_ens_col2 >= 0.5)
+
+  # reverse case, sym in first and id in second column
+  if (!is_ens[1] && is_ens[2]) return(SYM_IDS)
+
+  # regular cases. sum of booleans returns ints. convert to char to string match
+  feature_type <- switch(as.character(sum(is_ens)),
+                         "0" = SYM_SYM,
+                         "1" = IDS_SYM,
+                         "2" = IDS_IDS)
```

Comment:
I know this is ugly, I'm open to suggestions that don't involve 3 ifs

Reply:
given how `switch` works, the other alternative, also ugly, would be:

```suggestion
  # reverse case, sym in first and id in second column
  if (!is_ens[1] && is_ens[2]) return(SYM_IDS)

  # regular cases. sum of booleans returns ints. convert to char to string match
  feature_type <- switch(sum(is_ens) + 1,
                         SYM_SYM,
                         IDS_SYM,
                         IDS_IDS)
```

Reply:
Why don't you like this? I think it's a fine alternative to all the ifs, and is more readable than the sum of bools that I originally wrote